File: android/testSrc/com/android/tools/idea/run/AndroidProcessHandlerTest.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.run;

import com.android.ddmlib.Client;
import com.android.ddmlib.ClientData;
import com.android.ddmlib.IDevice;
import com.android.ddmlib.IShellOutputReceiver;
import com.android.sdklib.AndroidVersion;
import com.android.tools.idea.flags.StudioFlags;
import com.android.tools.idea.logcat.AndroidLogcatReceiver;
import com.android.tools.idea.logcat.AndroidLogcatService;
import com.android.tools.idea.logcat.output.LogcatOutputConfigurableProvider;
import com.android.tools.idea.logcat.output.LogcatOutputSettings;
import com.intellij.execution.process.ProcessAdapter;
import com.intellij.execution.process.ProcessEvent;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Disposer;
import com.intellij.openapi.util.Key;
import java.util.Collections;
import org.easymock.EasyMock;
import org.jetbrains.android.AndroidTestCase;
import org.jetbrains.annotations.NotNull;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import static com.google.common.truth.Truth.assertThat;

public class AndroidProcessHandlerTest extends AndroidTestCase {
  private static final String DEVICE_NAME = "myDevice";
  private static final int DEVICE_API_LEVEL = 25;
  private static final String DEVICE_SERIAL_NUMBER = "device-1";
  private static final String APPLICATION_ID = "FooApp";
  private static final int CLIENT_PID = 1493;

  private Project myProject;
  private AndroidProcessHandler myProcessHandler;
  private AndroidLogcatService myLogcatService;
  private CountDownLatch myExecuteShellCommandLatch;
  private CountDownLatch myFindClientLatch;
  private IDevice myDevice;
  private Client myClient;

  @Override
  protected void setUp() throws Exception {
    super.setUp();
    StudioFlags.RUNDEBUG_LOGCAT_CONSOLE_OUTPUT_ENABLED.override(true);

    myDevice = EasyMock.createMock(IDevice.class);
    EasyMock.expect(myDevice.getName()).andReturn(DEVICE_NAME).anyTimes();
    EasyMock.expect(myDevice.getVersion()).andReturn(new AndroidVersion(DEVICE_API_LEVEL)).anyTimes();
    EasyMock.expect(myDevice.getClientName(EasyMock.anyInt())).andReturn(APPLICATION_ID).anyTimes();
    EasyMock.expect(myDevice.getClient(APPLICATION_ID)).andAnswer(() -> myClient).anyTimes();
    EasyMock.expect(myDevice.getClients()).andAnswer(() -> new Client[]{myClient}).anyTimes();
    EasyMock.expect(myDevice.getSerialNumber()).andReturn(DEVICE_SERIAL_NUMBER).anyTimes();
    EasyMock.expect(myDevice.isOnline()).andReturn(true).anyTimes();

    myDevice.executeShellCommand(EasyMock.eq("logcat --help"), EasyMock.anyObject(), EasyMock.eq(10L), EasyMock.eq(TimeUnit.SECONDS));

    EasyMock.expectLastCall().andAnswer(() -> {
      sendTextLine((IShellOutputReceiver)EasyMock.getCurrentArguments()[1], "epoch");
      myExecuteShellCommandLatch.countDown();

      return null;
    }).anyTimes();

    myDevice.executeShellCommand(EasyMock.eq("logcat -v long -v epoch"), EasyMock.anyObject(), EasyMock.eq(0L),
                                 EasyMock.eq(TimeUnit.MILLISECONDS));

    EasyMock.expectLastCall().andAnswer(() -> {
      AndroidLogcatReceiver receiver = (AndroidLogcatReceiver)EasyMock.getCurrentArguments()[1];
      sendTextLine(receiver, "[ 1503099551.439 1493:1595 W/DummyFirst     ]");
      sendTextLine(receiver, "First Line1");
      sendTextLine(receiver, "First Line2");
      sendTextLine(receiver, "First Line3");
      sendTextLine(receiver, "[ 1505950751.439 1493:1595 W/DummySecond     ]");
      sendTextLine(receiver, "Second Line1");
      receiver.cancel();
      myExecuteShellCommandLatch.countDown();
      return null;
    }).anyTimes();

    ClientData clientData = EasyMock.createMock(ClientData.class);
    EasyMock.expect(clientData.getClientDescription()).andReturn(APPLICATION_ID).anyTimes();
    EasyMock.expect(clientData.getPackageName()).andReturn(APPLICATION_ID).anyTimes();
    EasyMock.expect(clientData.getPid()).andReturn(CLIENT_PID).anyTimes();

    myClient = EasyMock.createMock(Client.class);
    EasyMock.expect(myClient.getDevice()).andReturn(myDevice).anyTimes();
    EasyMock.expect(myClient.getClientData()).andReturn(clientData).anyTimes();
    EasyMock.expect(myClient.isValid()).andReturn(true).anyTimes();

    myFindClientLatch = new CountDownLatch(0);
    DeploymentApplicationService service = EasyMock.createMock(DeploymentApplicationService.class);
    EasyMock.expect(service.findClient(EasyMock.same(myDevice), EasyMock.same(APPLICATION_ID)))
      .andAnswer(() -> {
        if (myFindClientLatch.getCount() == 0) {
          return Collections.singletonList(myClient);
        }
        myFindClientLatch.countDown();
        return Collections.emptyList();
      }).anyTimes();

    CountDownLatch projectDisposedLatch = new CountDownLatch(1);
    myProject = EasyMock.createMock(Project.class);
    myProject.dispose();
    EasyMock.expectLastCall().andAnswer(() -> {
      projectDisposedLatch.countDown();
      return null;
    }).once();
    EasyMock.expect(myProject.isDisposed()).andAnswer(() -> projectDisposedLatch.getCount() == 0).anyTimes();
    EasyMock.expect(myProject.getDisposed()).andAnswer(() -> o -> myProject.isDisposed()).anyTimes();

    EasyMock.replay(myDevice, myClient, clientData, service, myProject);

    myLogcatService = AndroidLogcatService.getInstance();

    myProcessHandler = new AndroidProcessHandler.Builder(myProject)
      .setApplicationId(APPLICATION_ID)
      .setDeploymentApplicationService(service)
      .build();
  }

  @Override
  protected void tearDown() throws Exception {
    try {
      if (myLogcatService != null) {
        myLogcatService.shutdown();
      }
      StudioFlags.RUNDEBUG_LOGCAT_CONSOLE_OUTPUT_ENABLED.clearOverride();
      LogcatOutputSettings.getInstance().reset();
      Disposer.dispose(myProject);
      Disposer.dispose(DeploymentApplicationService.getInstance());
    }
    finally {
      super.tearDown();
    }
  }

  public void testLogcatMessagesAreForwardedAsProcessEvents() throws Exception {
    // Prepare
    myExecuteShellCommandLatch = new CountDownLatch(2);

    Collection<String> text = new ArrayList<>();
    myProcessHandler.addProcessListener(new ProcessAdapter() {
      @Override
      public void onTextAvailable(@NotNull ProcessEvent event, @NotNull Key outputType) {
        text.add(event.getText());
      }
    });

    // Act
    myProcessHandler.addTargetDevice(myDevice);
    myExecuteShellCommandLatch.await();

    assertThat(text).isEqualTo(Arrays.asList(
      "Connected to process 1493 on device 'myDevice'.\n",
      LogcatOutputConfigurableProvider.BANNER_MESSAGE + '\n',
      "W/DummyFirst: First Line1\n",
      "    First Line2\n",
      "    First Line3\n",
      "W/DummySecond: Second Line1\n"));

    myProcessHandler.deviceDisconnected(myDevice);
  }

  public void testLogcatMessagesAreForwardedWhenClientDelayed() throws Exception {
    // Prepare
    myExecuteShellCommandLatch = new CountDownLatch(2);
    myFindClientLatch = new CountDownLatch(1);

    Collection<String> text = new ArrayList<>();
    myProcessHandler.addProcessListener(new ProcessAdapter() {
      @Override
      public void onTextAvailable(@NotNull ProcessEvent event, @NotNull Key outputType) {
        text.add(event.getText());
      }
    });

    // Act
    myProcessHandler.addTargetDevice(myDevice);
    myExecuteShellCommandLatch.await();

    assertThat(text).isEqualTo(Arrays.asList(
      "Waiting for process to come online...\n",
      "Connected to process 1493 on device 'myDevice'.\n",
      LogcatOutputConfigurableProvider.BANNER_MESSAGE + '\n',
      "W/DummyFirst: First Line1\n",
      "    First Line2\n",
      "    First Line3\n",
      "W/DummySecond: Second Line1\n"));

    myProcessHandler.deviceDisconnected(myDevice);
  }

  public void testLogcatMessagesAreNotForwardedIfFeatureDisabled() throws Exception {
    // Prepare
    myExecuteShellCommandLatch = new CountDownLatch(0);
    StudioFlags.RUNDEBUG_LOGCAT_CONSOLE_OUTPUT_ENABLED.override(false);

    List<ProcessEvent> events = new ArrayList<>();
    myProcessHandler.addProcessListener(new ProcessAdapter() {
      @Override
      public void onTextAvailable(@NotNull ProcessEvent event, @NotNull Key outputType) {
        events.add(event);
      }
    });

    // Act
    myProcessHandler.addTargetDevice(myDevice);

    // Assert
    assertThat(events.size()).isEqualTo(1);
    assertThat(events.get(0).getText()).isEqualTo("Connected to process 1493 on device 'myDevice'.\n");
    myProcessHandler.deviceDisconnected(myDevice);
  }

  public void testLogcatMessagesAreNotForwardedIfSettingDisabled() throws Exception {
    // Prepare
    myExecuteShellCommandLatch = new CountDownLatch(0);

    LogcatOutputSettings.getInstance().setRunOutputEnabled(false);
    LogcatOutputSettings.getInstance().setDebugOutputEnabled(false);

    List<ProcessEvent> events = new ArrayList<>();
    myProcessHandler.addProcessListener(new ProcessAdapter() {
      @Override
      public void onTextAvailable(@NotNull ProcessEvent event, @NotNull Key outputType) {
        events.add(event);
      }
    });

    // Act
    myProcessHandler.addTargetDevice(myDevice);

    // Assert
    assertThat(events.size()).isEqualTo(1);
    assertThat(events.get(0).getText()).isEqualTo("Connected to process 1493 on device 'myDevice'.\n");

    myProcessHandler.deviceDisconnected(myDevice);
  }

  private static void sendTextLine(@NotNull IShellOutputReceiver receiver, @NotNull String s) {
    byte[] bytes = (s + "\n").getBytes(StandardCharsets.UTF_8);
    receiver.addOutput(bytes, 0, bytes.length);
  }
}

File: android/testSrc/com/android/tools/idea/run/AndroidProcessHandlerTest.kt
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.run

import com.android.ddmlib.IDevice
import com.android.sdklib.AndroidVersion
import com.android.testutils.MockitoKt.any
import com.android.testutils.MockitoKt.eq
import com.android.tools.idea.run.deployable.SwappableProcessHandler
import com.android.tools.idea.run.deployment.AndroidExecutionTarget
import com.google.common.truth.Truth.assertThat
import com.intellij.execution.ExecutionTarget
import com.intellij.execution.ExecutionTargetManager
import com.intellij.execution.process.AnsiEscapeDecoder
import com.intellij.execution.process.ProcessListener
import com.intellij.execution.process.ProcessOutputTypes
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Key
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.mockito.ArgumentMatchers.anyBoolean
import org.mockito.ArgumentMatchers.argThat
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.inOrder
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations.initMocks

/**
 * Unit test for [AndroidProcessHandler].
 */
@RunWith(JUnit4::class)
class AndroidProcessHandlerTest {
  companion object {
    const val TARGET_APP_NAME: String = "example.target.app"
  }

  @Mock lateinit var mockProject: Project
  @Mock lateinit var mockExecutionTargetManager: ExecutionTargetManager
  @Mock lateinit var mockExecutionTarget: AndroidExecutionTarget
  @Mock lateinit var mockDeploymentAppService: DeploymentApplicationService
  @Mock lateinit var mockMonitorManager: AndroidProcessMonitorManager
  @Mock lateinit var mockProcessListener: ProcessListener
  @Mock lateinit var mockAnsiEscapeDecoder: AnsiEscapeDecoder
  var captureLogcat: Boolean = true
  var autoTerminate: Boolean = true

  val handler: AndroidProcessHandler by lazy {
    AndroidProcessHandler(
      mockProject,
      TARGET_APP_NAME,
      captureLogcat,
      autoTerminate,
      mockAnsiEscapeDecoder,
      mockDeploymentAppService
    ) { _, _, emitter, listener ->
      textEmitter = emitter
      monitorManagerListener = listener
      mockMonitorManager
    }.apply {
      addProcessListener(mockProcessListener)
      startNotify()
    }
  }
  lateinit var textEmitter: TextEmitter
  lateinit var monitorManagerListener: AndroidProcessMonitorManagerListener

  @Before
  fun setUp() {
    initMocks(this)

    `when`(mockProject.getService(eq(ExecutionTargetManager::class.java)))
      .thenReturn(mockExecutionTargetManager)
    `when`(mockExecutionTargetManager.activeTarget).thenReturn(mockExecutionTarget)
    `when`(mockAnsiEscapeDecoder.escapeText(any(), any(), any())).then { invocation ->
      val (text, attributes, textAcceptor) = invocation.arguments
      text as String
      attributes as Key<*>
      textAcceptor as AnsiEscapeDecoder.ColoredTextAcceptor
      textAcceptor.coloredTextAvailable(text, attributes)
    }
  }

  @Test
  fun handlerIsRegisteredToCopyableUserData() {
    assertThat(handler.getCopyableUserData(SwappableProcessHandler.EXTENSION_KEY)).isSameAs(handler)
  }

  @Test
  fun runProcessOnOneDevice() {
    handler

    val inOrder = inOrder(mockProcessListener)
    inOrder.verify(mockProcessListener).startNotified(any())
    assertThat(handler.getUserData(AndroidSessionInfo.ANDROID_DEVICE_API_LEVEL)).isNull()

    val mockDevice = createMockDevice(28)
    handler.addTargetDevice(mockDevice)
    assertThat(handler.getUserData(AndroidSessionInfo.ANDROID_DEVICE_API_LEVEL)).isEqualTo(AndroidVersion(28))
    verify(mockMonitorManager).add(eq(mockDevice) ?: mockDevice)

    monitorManagerListener.onAllTargetProcessesTerminated()
    assertThat(handler.isProcessTerminating || handler.isProcessTerminated).isTrue()

    inOrder.verify(mockProcessListener).processWillTerminate(any(), /*willBeDestroyed=*/eq(true))
    inOrder.verify(mockProcessListener, timeout(1000)).processTerminated(any())
    inOrder.verifyNoMoreInteractions()

    assertThat(handler.isProcessTerminated).isTrue()
  }

  @Test
  fun runProcessOnMultipleDevices() {
    handler

    val inOrder = inOrder(mockProcessListener)
    inOrder.verify(mockProcessListener).startNotified(any())
    assertThat(handler.getUserData(AndroidSessionInfo.ANDROID_DEVICE_API_LEVEL)).isNull()

    handler.addTargetDevice(createMockDevice(28))
    assertThat(handler.getUserData(AndroidSessionInfo.ANDROID_DEVICE_API_LEVEL)).isEqualTo(AndroidVersion(28))

    handler.addTargetDevice(createMockDevice(27))
    // AndroidSessionInfo.ANDROID_DEVICE_API_LEVEL should return the lowest API level among all managed devices.
    assertThat(handler.getUserData(AndroidSessionInfo.ANDROID_DEVICE_API_LEVEL)).isEqualTo(AndroidVersion(27))

    handler.addTargetDevice(createMockDevice(29))
    assertThat(handler.getUserData(AndroidSessionInfo.ANDROID_DEVICE_API_LEVEL)).isEqualTo(AndroidVersion(27))

    monitorManagerListener.onAllTargetProcessesTerminated()
    assertThat(handler.isProcessTerminating || handler.isProcessTerminated).isTrue()

    inOrder.verify(mockProcessListener).processWillTerminate(any(), /*willBeDestroyed=*/eq(true))
    inOrder.verify(mockProcessListener, timeout(1000)).processTerminated(any())
    inOrder.verifyNoMoreInteractions()

    assertThat(handler.isProcessTerminated).isTrue()
  }

  @Test
  fun textEmitterShouldRedirectToNotifyText() {
    handler
    textEmitter.emit("test emit message", ProcessOutputTypes.STDOUT)
    verify(mockProcessListener).onTextAvailable(argThat { event -> event.text == "test emit message" }, eq(ProcessOutputTypes.STDOUT))
  }

  @Test
  fun destroyProcess() {
    val inOrder = inOrder(mockProcessListener)

    handler.destroyProcess()
    assertThat(handler.isProcessTerminating || handler.isProcessTerminated).isTrue()

    inOrder.verify(mockProcessListener).startNotified(any())
    inOrder.verify(mockProcessListener).processWillTerminate(any(), /*willBeDestroyed=*/eq(true))
    inOrder.verify(mockProcessListener, timeout(1000)).processTerminated(any())
    inOrder.verifyNoMoreInteractions()

    assertThat(handler.isProcessTerminated).isTrue()
  }

  @Test
  fun detachProcess() {
    val inOrder = inOrder(mockProcessListener)

    handler.detachProcess()
    assertThat(handler.isProcessTerminating || handler.isProcessTerminated).isTrue()

    inOrder.verify(mockProcessListener).startNotified(any())
    inOrder.verify(mockProcessListener).processWillTerminate(any(), /*willBeDestroyed=*/eq(false))
    inOrder.verify(mockProcessListener, timeout(1000)).processTerminated(any())
    inOrder.verifyNoMoreInteractions()

    assertThat(handler.isProcessTerminated).isTrue()
  }

  @Test
  fun canKillProcess_returnsFalseWhenNoAssociatedDevices() {
    assertThat(handler.canKillProcess()).isFalse()
  }

  @Test
  fun canKillProcess_returnsTrueWhenThereIsAnyAssociatedDevice() {
    val nonAssociatedDevice = mock(IDevice::class.java)
    val associatedDevice = mock(IDevice::class.java)

    `when`(mockExecutionTarget.runningDevices).thenReturn(listOf(nonAssociatedDevice, associatedDevice))
    `when`(mockMonitorManager.isAssociated(associatedDevice)).thenReturn(true)

    assertThat(handler.canKillProcess()).isTrue()
  }

  @Test
  fun canKillProcess_returnsFalseWhenThereAreNoAssociatedDevices() {
    val nonAssociatedDevice1 = mock(IDevice::class.java)
    val nonAssociatedDevice2 = mock(IDevice::class.java)

    `when`(mockExecutionTarget.runningDevices).thenReturn(listOf(nonAssociatedDevice1, nonAssociatedDevice2))

    assertThat(handler.canKillProcess()).isFalse()
  }

  @Test
  fun canKillProcess_returnsFalseWhenActiveTargetIsNotAndroidTarget() {
    `when`(mockExecutionTargetManager.activeTarget).thenReturn(mock(ExecutionTarget::class.java))

    assertThat(handler.canKillProcess()).isFalse()
  }

  @Test
  fun ProcessHandlerShouldAutoTerminateWhenAutoTerminateIsEnabled() {
    autoTerminate = true

    handler.addTargetDevice(createMockDevice(28))
    monitorManagerListener.onAllTargetProcessesTerminated()

    assertThat(handler.isProcessTerminating || handler.isProcessTerminated).isTrue()
    inOrder(mockProcessListener).apply {
      verify(mockProcessListener).processWillTerminate(any(), /*willBeDestroyed=*/eq(true))
      verify(mockProcessListener, timeout(1000)).processTerminated(any())
      verifyNoMoreInteractions()
    }

    assertThat(handler.isProcessTerminated).isTrue()
  }

  @Test
  fun ProcessHandlerShouldNotAutoTerminateWhenAutoTerminateIsOff() {
    autoTerminate = false

    handler.addTargetDevice(createMockDevice(28))
    monitorManagerListener.onAllTargetProcessesTerminated()
    assertThat(handler.isProcessTerminating || handler.isProcessTerminated).isFalse()

    verify(mockProcessListener, never()).processWillTerminate(any(), anyBoolean())
    assertThat(handler.isProcessTerminated).isFalse()
  }

  private fun createMockDevice(apiVersion: Int): IDevice {
    val mockDevice = mock(IDevice::class.java)
    `when`(mockDevice.version).thenReturn(AndroidVersion(apiVersion))
    return mockDevice
  }
}